addons/mail/controllers/im_status.py

```diff
@@ -0,0 +1,30 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import http
+from odoo.http import request
+
+from datetime import datetime, timedelta, time
+
+
+class ImStatusController(http.Controller):
+    @http.route("/mail/im_status", methods=["POST"], type="json", auth="user")
+    def set_im_status(self, action):
+        user = request.env.user
+        user.forced_im_status = None if action == "online" else action
+        user._bus_send("bus.bus/im_status_updated", {"im_status": action, "partner_id": user.partner_id.id}, subchannel="presence")
+
+    @http.route("/mail/custom_status", methods=["POST"], type="json", auth="user")
+    def set_custom_status(self, custom_status, reset_after):
+        user = request.env.user
+        user.custom_status = custom_status
+        match reset_after:
+            case "today":
+                user.reset_custom_status_datetime = datetime.combine(datetime.utcnow().date() + timedelta(days=1), time())
+            case "1_hour":
+                user.reset_custom_status_datetime = datetime.utcnow() + timedelta(hours=1)
+            case "4_hour":
+                user.reset_custom_status_datetime = datetime.utcnow() + timedelta(hours=4)
+            case "never":
+                user.reset_custom_status_datetime = None
```

Comment:
I believe it would only move the logic in another part because the `reset_after` parameter comes from a html selection element so a switch case is bound to happen somewhere.

Also the `today` and `never` options need a specific logic anyway so it wouldn't remove much code.

Reply:
I don't really agree, no switch case is needed, you'll just need to change the options values to `1`/`4`/`false`. You'll send to the server the exact values it needs and avoid this processing all-together. `never` does not look like a special case to me, it's the absence of value. And about today, why omitting the time? Indeed `today` is a bit trickier, but it can very much be deduced with luxon.